### DISCUSSION TITLE
Discussion: Tables

### DISCUSSION THREAD

**Discussion by sturobson** [visual-framework/vf-core#1474] [General]
I'll add more words here soon...

**Comment by gabsie** [comment]
Hi, everyone

I am kind of after these types of tables, and I will enumerate the functionalities needed
https://medium.com/design-with-figma/the-ultimate-guide-to-designing-data-tables-7db29713a85a

1. Tables will need to fit thousands of entries
2. I need statuses  - Valid/Invalid/Validating, traffic light system would be great
3. I need the ability to edit an individual entry
4. I need an ability to search and filter only a couple of results

Will add more here :)
Gabs

**Comment by khawkins98** [comment]
There's almost certainly many separate sets of work to be done, ideas:

- High-level documentation on how to do tables
- Adding in responsive layout
- Scrolling tables
- ... and all the things @gabsie has mentioned above

So I think it's worth adding Tables as a milestone at https://github.com/visual-framework/vf-core/milestones to be worked on over the next few months. 

Once we have the tasks, we'll then be able to work on them over time.

**Comment by khawkins98** [comment]
@gabsie shard this highly interactive example from NCBI 

https://www.ncbi.nlm.nih.gov/labs/virus/vssi/#/virus?SeqType_s=Nucleotide

![image](https://user-images.githubusercontent.com/928100/113733165-81497f00-96fa-11eb-81e9-5c1425e103ee.png)

**Comment by sturobson** [comment]
There's the question of how much JavaScript comes from `vf-core` considering the `vf-table` will undoubtedly be used alongside JS frameworks … and from experience - they can easily blow up if the JS isn't written 'their way'. 

This is why `vf-form` things come with no JS for validation.

**Comment by khawkins98** [comment]
Was chatting with @sturobson and showed a bunch of common use cases, will share them here:

### Ensembl

http://www.ensembl.org/Homo_sapiens/Gene/Matches?db=core;g=ENSG00000139618;r=13:32315086-32400268

![image](https://user-images.githubusercontent.com/928100/113999306-36df1400-985a-11eb-80ba-23cca3d6b397.png)

### HMMER

https://www.ebi.ac.uk/Tools/hmmer/results/87CED718-9848-11EB-94EC-3FBFF75AEC3D/score

![image](https://user-images.githubusercontent.com/928100/113999376-48282080-985a-11eb-910b-1dcae2350cdb.png)


### Expression Atlas

https://www.ebi.ac.uk/gxa/experiments?experimentalFactors=age

![image](https://user-images.githubusercontent.com/928100/113999428-5413e280-985a-11eb-9677-9e9d2fda5bb1.png)

### ENA

https://www.ebi.ac.uk/ena/browser/view/PRJEB12455

![image](https://user-images.githubusercontent.com/928100/114000067-ef0cbc80-985a-11eb-8b32-e3b0940f62ea.png)

### Chembl

https://www.ebi.ac.uk/chembl/g/#search_results/compounds

![image](https://user-images.githubusercontent.com/928100/114000211-1499c600-985b-11eb-9004-b32e63c2af77.png)

### UniProt

https://www.uniprot.org/uniparc/UPI0000000001?sort=score

![image](https://user-images.githubusercontent.com/928100/114000171-0b105e00-985b-11eb-8ec6-e984a55422e0.png)



### EBI News

And lest we forget, boring old content tables

https://www.ebi.ac.uk/about/news/press-releases/embl-ebi-launches-covid-19-data-portal

![image](https://user-images.githubusercontent.com/928100/113999894-c1c00e80-985a-11eb-905b-a62f7ac689f6.png)

**Comment by khawkins98** [comment]
Also, a vanilla JS table sorter: https://tofsjonas.github.io/sortable/

**Comment by cindyebi** [comment]
@khawkins98 @sturobson I noticed that we have two action placements on the Kitchen Sink variant, one on the right-hand column and the other one at the bottom row. Are there an instance of when they both appear at the same time? If not, would it be better if we choose one? 
![Screenshot 2021-04-08 at 14 36 46](https://user-images.githubusercontent.com/75834903/114036303-dc54b080-9877-11eb-9b0e-cc8c38d26a38.png)

Also, it would be nice to know whenin what context people should be using the "Has Footer" variant as well. 
![Screenshot 2021-04-08 at 14 40 22](https://user-images.githubusercontent.com/75834903/114036886-5ab15280-9878-11eb-892e-9923cd78b459.png)

**Reply by khawkins98** [reply]
I don't think either of these variants has seen much adoption.

As for the right side vs footer actions, I think the context is different. Where the in-column action applies to a singular row, and the footer applies to all selected rows. I wouldn't expect that the use case is the same. 

I think working through some real-world examples would be more effective than lorem ipsum.

**Comment by cindyebi** [comment]
**Sortable table variant**
There was a discussion around the sortable variant's icon and behaviour on one of the vf catch-ups last week. 
From an accessibility point of view, I think that only showing the sorting icons on hover isn't good for discoverability, but having all of the icons present by default is probably not the right approach either. @gabsie let me know if you have any thoughts on this. 

![Screenshot 2021-04-19 at 10 41 57](https://user-images.githubusercontent.com/75834903/115223304-8e576c80-a103-11eb-81f4-b69656303f17.png)

I attached here some examples as references.
[Material UI](https://material-ui.com/components/tables/)
Material design system shows a column sorted by default, e.g."Calories" as only one column could be sorted at a time.
<img width="802" alt="Screenshot 2021-04-19 at 10 44 03" src="https://user-images.githubusercontent.com/75834903/115224414-babfb880-a104-11eb-84b1-5c4b9d3a82e2.png">

[Carbon Design System](https://www.carbondesignsystem.com/components/data-table/usage/) shows the sorting function on mouse hover with a highlighted background. 

![Screenshot 2021-04-19 at 10 42 28](https://user-images.githubusercontent.com/75834903/115224971-77b21500-a105-11eb-8c15-cca2bda7c4f3.png)

The 'active' column is represented with a blue border:
<img width="891" alt="Screenshot 2021-04-19 at 10 43 13" src="https://user-images.githubusercontent.com/75834903/115225279-d081ad80-a105-11eb-9c04-8168294efa3b.png">

Another example:
![b2ccfbe06a08ff22f06b1f7440de2aa4a568e3c0](https://user-images.githubusercontent.com/75834903/115225875-80571b00-a106-11eb-8306-f3677fa13e43.png)

I prefer this version of where all icons are present in each column as a _signifier_ that the column is sortable with one column sorted and highlighted by default.

**Comment by cindyebi** [comment]
Additionally, there was also a discussion around the **Selectable Selected** variant last week. At the moment the action links are located at the bottom of the table, which will work fine on a small table that doesn't have many rows. However, it would be difficult to find or reach when the table has many rows as the users will need to scroll down to the bottom of the page to find the action items.  
![Screenshot 2021-04-20 at 14 41 41](https://user-images.githubusercontent.com/75834903/115405851-86bcc400-a1e6-11eb-8417-1ee73d8ef5d3.png)

An example from [Carbon Design System](https://www.carbondesignsystem.com/components/data-table/usage/) that shows the action items within the table header: 
![datatable_search](https://user-images.githubusercontent.com/75834903/115406405-0a76b080-a1e7-11eb-8b04-c2e87618d16a.gif)

[Material UI](https://material-ui.com/components/tables/) also shows the action items on the header or top part of the table:
![Screenshot 2021-04-20 at 14 47 50](https://user-images.githubusercontent.com/75834903/115407454-f2ebf780-a1e7-11eb-9165-b76971d13841.png)

Eventually, it is also helpful to have a 'fixed' header where it stays in place when scrolled. 
![1_yEYBLf6pb6c8VnCCTtpRcw](https://user-images.githubusercontent.com/75834903/115407328-d780ec80-a1e7-11eb-9810-41d3f1fb6756.gif)
